core/src/com/google/inject/internal/LinkageContext.java

```diff
@@ -58,30 +60,31 @@ final class LinkageContext {
    * @return a method handle that will invoke the given factory, resolving cycles as needed, using
    *     the {@link InternalFactory#FACTORY_TYPE} signature.
    */
-  MethodHandle makeHandle(InternalFactory<?> source, Supplier<MethodHandle> factory) {
-    var previous = linkingFactories.putIfAbsent(source, CONSTRUCTING);
+  MethodHandleResult makeHandle(InternalFactory<?> factory, boolean linked) {
+    var previous = linkingFactories.putIfAbsent(factory, CONSTRUCTING);
     if (previous == CONSTRUCTING) {
       // We are the first to 're-enter' this factory, so we need to create a MutableCallSite that
       // will be used to resolve the cycle.
       previous = new MutableCallSite(InternalMethodHandles.FACTORY_TYPE);
-      linkingFactories.put(source, previous);
+      linkingFactories.put(factory, previous);
     }
     if (previous instanceof MutableCallSite) {
       // We are re-entering the same factory, so we can just return the dynamic invoker and rely on
       // the original invocation to finish the construction and call setTarget to finalize the
       // callsite.
-      return ((MutableCallSite) previous).dynamicInvoker();
+      return InternalFactory.makeUncachable(((MutableCallSite) previous).dynamicInvoker());
```

Comment:
Can you elaborate in comments on why this is uncachable?

Reply:
Added a comment.  There is no correct caching setting here so we don't try, when the original invocation returns we will cache that result.